### DISCUSSION TITLE
How would one create roles, add users to roles, modify tables with roles, etc. with supabase-js?

### DISCUSSION THREAD

**Discussion by jake-edgenaut** [supabase/supabase#1701] [Questions] [answered]
Would you have to create a stored procedure and run it with rpc()? Apologies if I missed a major part of the documentation, but didn't see a way to do this with supabase-js. TIA

**Comment by prescience-data** [comment]
I think the idea is that you put all that stuff in your database schema with `GRANT`, `CREATE POLICY`, etc...

I found the original [PostgREST](https://github.com/PostgREST/postgrest) documentation has a lot more info/guidance while Supabase is building out their docs: https://postgrest.org/en/stable/

**Comment by roryw10** [comment, marked as answer]
These things can either be achieved using the SQL editor or connecting to the database directly. Another option could be to call via the RPC function.

If you want to do DDL via the rest API, you can check this out >> https://github.com/supabase/postgres-meta

**Reply by jake-edgenaut** [reply]
Hi Rory, I should've clarified in my question that this is for admin users of my app to assign roles to users, change privileges, etc. So the editor is out and so is postgres-meta as it is not secured and meant for back-end work. So I think I'm limited to RPC via supabase-js or perhaps PostgREST via postgrest-js--it's not clear whether the latter is secure or not from the docs. The PostgREST docs talk about authenticating API requests with JWT tokens.

**Reply by steve-chavez** [reply]
@ornithophile Note that `supabase-js` has [postgrest-js](https://github.com/supabase/supabase-js/blob/master/package.json#L40) as a dependency, so you only need to look at `supabase-js`.

> it's not clear whether the latter is secure or not from the docs.

It should be safe to use rpc for role management, as long as you restrict calling the function to high-privilege roles.

You could do so by revoking the default execute privilege for the function and then grant it only to `admin` roles.

```sql
revoke all privileges on function manage_role(text) from public;
grant execute on function manage_role(text) to admin; -- assuming you have an "admin" role
```

(You should check that calling that function through rpc doesn't work for `anon` or other low-privilege roles)

Another option, if you don't want to create another role, would be to have something like an `is_admin` column on your users table. Then do a check inside the function for this column based on `auth.uid()` and raise an exception if it's called by non-admins.

**Reply by jake-edgenaut** [reply]
Thanks @steve-chavez !
